File: practical-ios-concurrency/5-practical-ios-concurrency-m5-subclass-exercise-files/demos/after/TicketSim/ViewController.m
//
//  ViewController.m
//  TicketSim
//
//  Created by Brice Wilson on 5/2014.
//  Copyright (c) 2014 Brice Wilson. All rights reserved.
//

#import "ViewController.h"
#import "Simulator.h"

@interface ViewController () {
    
    NSArray *customers;
    int currentCustomerIndex;
    NSOperationQueue *ticketQueue;
}

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
    
    customers = [NSArray arrayWithObjects:@"Audrey", @"Bub", @"Leigh Ann", @"Darshan", @"Evan", nil];
    currentCustomerIndex = 0;
    [[self customerNameLabel] setText:[customers objectAtIndex:currentCustomerIndex]];
    ticketQueue = [[NSOperationQueue alloc] init];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)buyTicketsClicked:(id)sender {
    
    NSString *currentCustomerName = [customers objectAtIndex:currentCustomerIndex];
    
    BuyTicketsOperation *buyTicketsOp = [[BuyTicketsOperation alloc] initWithCustomerName:currentCustomerName];
    [buyTicketsOp setDelegate:self];

    NSBlockOperation *payOp = [NSBlockOperation blockOperationWithBlock:^{
        
        double time = [Simulator runSimulationWithMinTime:4 maxTime:10];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            
            [self logResult:[NSString stringWithFormat:@"%@ paid in %.02f seconds.", currentCustomerName, time]];
            
        });
        
    }];

    [buyTicketsOp addObserver:self
                   forKeyPath:@"isCancelled"
                      options:NSKeyValueObservingOptionNew
                      context:(__bridge void *)(currentCustomerName)];
    

    [payOp addDependency:buyTicketsOp];

    [ticketQueue addOperation:payOp];
    [ticketQueue addOperation:buyTicketsOp];
    
    
    if (currentCustomerIndex == [customers count] - 1) {
        [[self customerNameLabel] setText:@"No more customers"];
        [[self buyTicketsButton] setEnabled:NO];
    } else {
        [[self customerNameLabel] setText:[customers objectAtIndex:++currentCustomerIndex]];
    }
}

- (void)displayBuyTicketsResult:(SimulationResult *)result {
    
    NSString *outputString = [NSString stringWithFormat:@"Bought for %@ in %.02f seconds.", [result customerName], [result simulationTime]];
    [self logResult:outputString];
    
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context {
    
    
    if ([keyPath isEqualToString:@"isCancelled"]) {
        
        NSLog(@"Buying operation for %@ is cancelled", context);

    }
}


- (IBAction)resetClicked:(id)sender {
    [[self outputTextView] setText:@""];
    currentCustomerIndex = 0;
    [[self customerNameLabel] setText:[customers objectAtIndex:currentCustomerIndex]];
    [[self buyTicketsButton] setEnabled:YES];
}

- (IBAction)alphaChanged:(id)sender {
    UIColor *currentColor = [[self view] backgroundColor];
    [[self view] setBackgroundColor:[currentColor colorWithAlphaComponent:[[self alphaSlider] value]]];
}


- (IBAction)cancelClicked:(id)sender {
    
    [ticketQueue cancelAllOperations];
    
}

- (void)logResult:(NSString *)message {
    
    NSString *contents = [[NSString alloc] init];
    
    if ([[self outputTextView] hasText]) {
        contents = [[[self outputTextView] text] stringByAppendingString:@"\n"];
    }
    
	contents = [contents stringByAppendingString:message];
	[[self outputTextView] setText:contents];
    
}

@end

File: design-patterns-swift-behavioral/03/demos/demos/ResponderChain/ResponderChain/ViewController.swift
//
//  Design Pattern in Swift: Behavioral - https://app.pluralsight.com/library/courses/design-patterns-swift-behavioral
//
//  ViewController.swift
//  ResponderChain
//
//  Created by Nyisztor, Karoly on 8/14/17.
//  Copyright © 2017 Nyisztor, Karoly. All rights reserved.
//

import UIKit

extension UIView {
    override open func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        print("UIView: \(#function) called")
        next?.touchesBegan(touches, with: event)
    }
}

extension UIWindow {
    override open func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        print("UIWindow: \(#function) called")
        next?.touchesBegan(touches, with: event)
    }
}

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        print("ViewController: \(#function) called")
        next?.touchesBegan(touches, with: event)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}

